src/GafferImage/ImageWriter.cpp

```diff
@@ -1675,59 +1775,150 @@ void ImageWriter::execute() const
 		boost::filesystem::create_directories( directory );
 	}
 
-	if ( out->open( fileName, spec ) )
-	{
-		IECore::msg( IECore::MessageHandler::Info, this->relativeName( this->scriptNode() ), "Writing " + fileName );
-	}
-	else
-	{
-		throw IECore::Exception( boost::str( boost::format( "Could not open \"%s\", error = %s" ) % fileName % out->geterror() ) );
-	}
-
 	// Write out the channel data
 
-	const ColorSpace *appropriateColorSpaceNode = hasAlpha ? colorSpaceUnpremultedNode() : colorSpaceNode();
-
 	const Imath::Box2i extImageDataWindow( Imath::V2i( spec.x, spec.y ), Imath::V2i( spec.x + spec.width - 1, spec.y + spec.height - 1 ) );
 	const Imath::Box2i imageDataWindow( imageFormat.fromEXRSpace( extImageDataWindow ) );
 	const Imath::Box2i processDataWindow( BufferAlgo::intersection( imageDataWindow, dataWindow ) );
 
-	if( !deep )
+	struct Part
 	{
-		TileChannelDataProcessor processor;
+		std::string name;
+		std::string view;
+		std::vector< std::string > channels;
+		std::vector< std::string > channelNames;
+	};
+
+	std::vector< Part > parts;
 
-		if ( spec.tile_width == 0 )
+	std::string currentView = "";
+
+	for( const string &i : channelsToWrite )
+	{
+		auto && [ viewName, partName, channelName ] = viewPartChannelName( this, currentView, i );
+
+		size_t partIndex = std::distance(
+			parts.begin(),
+			std::find_if( parts.begin(), parts.end(), [partName] (Part const& p) { return p.name == partName; } )
+		);
+
+		if( partIndex >= parts.size() )
 		{
-			FlatScanlineWriter flatScanlineWriter( out, fileName, processDataWindow, imageFormat );
-			ImageAlgo::parallelGatherTiles( appropriateColorSpaceNode->outPlug(), spec.channelnames, processor, flatScanlineWriter, processDataWindow, ImageAlgo::TopToBottom );
-			flatScanlineWriter.finish();
+			parts.push_back( { partName, viewName, {}, {} } );
 		}
 		else
 		{
-			FlatTileWriter flatTileWriter( out, fileName, processDataWindow, imageFormat );
-			ImageAlgo::parallelGatherTiles( appropriateColorSpaceNode->outPlug(), spec.channelnames, processor, flatTileWriter, processDataWindow, ImageAlgo::TopToBottom );
-			flatTileWriter.finish();
+			if( parts[partIndex].view != viewName )
+			{
+				throw IECore::Exception( boost::str( boost::format( "Cannot write views \"%s\" and \"%s\" both to image part \"%s\"" ) % parts[partIndex].view % viewName % partName ) );
+			}
 		}
+		
+		parts[ partIndex ].channels.push_back( i );
+		parts[ partIndex ].channelNames.push_back( channelName );
+	}
 
+	if( parts.size() > 1 && !out->supports( "multiimage" ) )
+	{
+		throw IECore::Exception( boost::str( boost::format( "Cannot write multiple parts to this image format: \"%s\"" ) % fileName ) );
 	}
-	else
+
+	bool hasAlpha = false;
+	std::vector< ImageSpec > specs;
+	for( size_t partIndex = 0; partIndex < parts.size(); partIndex++ )
```

Comment:
We don't seem to need the numeric value of `partIndex` in the loop, so range-for would make things a bit more readable (`part` vs `parts[partIndex]` everywhere).